File: benchmarks/tasks/mandelbrot/mandelbrot.go
// Data parallelism: a 2000x2000 grid of independent float work, split between
// goroutines by taking every Nth row.

package main

import (
	"fmt"
	"runtime"
	"sync"
)

const side = 2000
const limit = 100

func band(start, step int64) int64 {
	var total int64
	for j := start; j < side; j += step {
		cy := float64(j)/float64(side)*3.0 - 1.5
		for i := int64(0); i < side; i++ {
			cx := float64(i)/float64(side)*3.0 - 2.0
			x, y := 0.0, 0.0
			var n int64
			for n < limit && x*x+y*y <= 4.0 {
				x, y = x*x-y*y+cx, 2.0*x*y+cy
				n++
			}
			total += n
		}
	}
	return total
}

func main() {
	workers := int64(runtime.NumCPU())
	sums := make([]int64, workers)
	var wg sync.WaitGroup
	for w := int64(0); w < workers; w++ {
		wg.Add(1)
		go func(w int64) {
			defer wg.Done()
			sums[w] = band(w, workers)
		}(w)
	}
	wg.Wait()
	var total int64
	for _, s := range sums {
		total += s
	}
	fmt.Println(total)
}

File: benchmarks/tasks/mandelbrot/mandelbrot.c
/* Data parallelism: a 2000x2000 grid of independent float work, split between
 * threads by taking every Nth row. */

#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <pthread.h>
#include <unistd.h>

#define SIDE 2000
#define LIMIT 100

typedef struct { int64_t start, step, total; } Band;

static void *band(void *arg) {
    Band *b = arg;
    int64_t total = 0;
    for (int64_t j = b->start; j < SIDE; j += b->step) {
        double cy = (double)j / (double)SIDE * 3.0 - 1.5;
        for (int64_t i = 0; i < SIDE; i++) {
            double cx = (double)i / (double)SIDE * 3.0 - 2.0;
            double x = 0.0, y = 0.0;
            int64_t n = 0;
            while (n < LIMIT && x * x + y * y <= 4.0) {
                double nx = x * x - y * y + cx;
                y = 2.0 * x * y + cy;
                x = nx;
                n++;
            }
            total += n;
        }
    }
    b->total = total;
    return NULL;
}

int main(void) {
    long workers = sysconf(_SC_NPROCESSORS_ONLN);
    if (workers < 1) workers = 4;
    pthread_t *ts = malloc(workers * sizeof(pthread_t));
    Band *bands = malloc(workers * sizeof(Band));
    for (long w = 0; w < workers; w++) {
        bands[w] = (Band){w, workers, 0};
        pthread_create(&ts[w], NULL, band, &bands[w]);
    }
    int64_t total = 0;
    for (long w = 0; w < workers; w++) {
        pthread_join(ts[w], NULL);
        total += bands[w].total;
    }
    printf("%lld\n", (long long)total);
    free(ts); free(bands);
    return 0;
}
